### ISSUE TITLE
Remove the web installer in favor of setting up via cli commands

### ISSUE THREAD

**Issue by u2mejc** [dokku/dokku#2247]
Description of problem:
- Default install of dokku exposes an insecure dokku admin access setup page on all interfaces.

How was dokku installed?:
- boostrap.sh

How reproducible:
- Fresh bootstrap.sh install

Additional info:
- From http://dokku.viewdocs.io/dokku/installation/:

> If you don't complete setup via the web installer (even if you set up SSH keys and virtual hosts otherwise) your dokku installation will remain vulnerable to anyone finding the setup page and inserting their key.

Allot of users use dokku on public facing VMs, the default install doesn't listen for a vhost and instead responds to any request for `/` on port 80. Should we add a secure option for the "attended" installs? Some new users ~~may~~ appear not to be aware if they haven't pushed an app yet and are only interacting with the CLI.

Possible changes:
- Add scary warning to CLI, everytime dokku is execute from CLI
- Randomly generated password on CLI for basic http auth during install (reset with a reinstall)
- Add CLI prompts for key / vhost settings (CLI or with the Ubuntu `dialog` command)
- Don't start the webserver until it's activated by a command inside the dokku CLI
- Add documentation around not starting dokku setup and instead document how to manually add ssh key / set hostname preference.

**Comment by josegonzalez** [comment]
👍 

@u2mejc want to start on a PR for these set of changes?

**Comment by u2mejc** [comment]
@josegonzalez happy to, which ones would you like to see in it?

**Comment by josegonzalez** [comment]
> Add scary warning to CLI, everytime dokku is execute from CLI
> Add CLI prompts for key / vhost settings (CLI or with the Ubuntu dialog command)


@michaelshobbs thoughts on dropping the web installer completely and making people set it up on the cli?

**Comment by michaelshobbs** [comment]
I'm 👍 on dropping the web installer and wrapping setup in a CLI

**Comment by josegonzalez** [comment]
I wonder what the backlash will be on doing that...

**Comment by u2mejc** [comment]
> I wonder what the backlash will be on doing that...

I believe non-existent. Dokku is a cli tool, but for some reason is includes like two settings you have to switch away from to cli into a website? If there is backlash, I expect it to be from the camp that says this is taking us further away from having a web gui someday, but I disagree that this would even be a building stone. I'll whip up a PR.

**Comment by u2mejc** [comment]
Checklist till completion: 
- [x] Update sshcommand to support listing - PR: https://github.com/dokku/sshcommand/pull/29
- [ ] Create ssh-keys plugin to manage keys - PR: https://github.com/dokku/dokku/pull/2287
- [ ] Ensure the deb prompts during interactive setup, add warning to CLI if authorized_keyfile is missing
  - should we add an option to disable for those who don't push remotly (like image taggers)?
- [ ] Purge all references to python setup website

**Comment by josegonzalez** [comment]
@andrewsomething How do you think this will impact DigitalOcean users?

**Comment by josegonzalez** [comment]
I really like what Gitlab did here (maybe it wasn't Gitlab but DigitalOcean):

They call the following in the `/root/.bashrc` (we can probably get away with something in `/etc/profile.d`:

```
bash /var/lib/digitalocean/first_login.sh
```

This ends up setting up the following:

![https://cl.ly/0i3Y47042M1J](https://d3vv6lp55qjaqc.cloudfront.net/items/3j0A2w1b111o0x37210X/Image%202017-01-22%20at%203.38.58%20PM.png?X-CloudApp-Visitor-Id=87115)

They also setup an motd:

```
cat /etc/update-motd.d/99-one-click
#!/bin/sh
#
# Configured as part of the DigitalOcean 1-Click Image build process

myip=$(hostname -I | awk '{print$1}')
cat <<EOF
-------------------------------------------------------------------------------
Thank you for using DigitalOcean's GitLab Application
Your GitLab instance can be accessed at http://$myip

You can find more information on using this image at: http://do.co/gitlabapp

Let's Encrypt has been pre-installed for you. If you have a domain name, and
you will be using it with this 1-Click app, please see: http://do.co/le-gitlab

-------------------------------------------------------------------------------
To delete this message of the day: rm -rf $(readlink -f ${0})
EOF
```

Then on ssh, you get the following prompt:

![https://cl.ly/3n2k0l1s1s0X](https://d3vv6lp55qjaqc.cloudfront.net/items/2t2l413H0U0B14181q1r/Image%202017-01-22%20at%203.39.21%20PM.png?X-CloudApp-Visitor-Id=87115)

Which, if you ctrl+c out of it, will result in this:

![https://cl.ly/1z2e0x1g3P23](https://d3vv6lp55qjaqc.cloudfront.net/items/2s3K2c3P3g2H0m0g2X3G/Image%202017-01-22%20at%203.40.49%20PM.png?X-CloudApp-Visitor-Id=87115)

We can probably replace our web panel with a similar workflow.

**Comment by u2mejc** [comment]
👍  I like it the idea of s simple splash page (ala apache / nginx), with no unauthorized ability to set ssh keys. Since we don't currently control the instances (I think we'll choose creating a container instead), the first login may have already happened to get Dokku installed. The single click apps we should probably be left to the cloud providers to maintain, too many clouds and unique setups for three of us to manage.

What I would do on a first run of dokku setup, once the setup is complete, is write a file: `~dokku/DokkuSetupComplete`. Removing that file would prompt Dokku to run setup again on next run.

**Comment by josegonzalez** [comment]
I don't want to litter the home directory anymore, but I get the idea. Can you screw around with a test DO instance for gitlab and figure out exactly how they did this?

**Comment by josegonzalez** [comment]
This is appended to the end of the root user login:

```
# Execute the first login script

bash /var/lib/digitalocean/first_login.sh
```

And here is the contents of that file:

```
#!/bin/bash
## PART: reconfigure upon root login
##
## vi: syntax=sh expandtab ts=4
#

myip=$(hostname -I |awk '{print$1}')
cat <<EOC
-----------------------------------------------------------------------------

GitLab is not configured. If you would like customize it yourself,
press 'n' now.

EOC

read -p "Okay to Configure GitLab (Y|n): " yesno

if [[ ! "${yesno^}" =~ "N" ]]; then


    mkdir -p /var/log/gitlab/reconfigure
    cp /etc/gitlab/gitlab.rb.digitalocean /etc/gitlab/gitlab.rb

    read -p "External URL (defaults to http://${myip}): " external_url
    echo "external_url \"${external_url:-http://${myip}}\"" >> /etc/gitlab/gitlab.rb

    checked=""
    until [[ "${checked}"  =~ "OK" ]];
    do
      read -p "Enter the GitLab 'root' user password: " gitlab_pass
      checked=$(/usr/sbin/cracklib-check <<<$gitlab_pass)
      echo "${checked}"
      checked=$(echo $checked | awk "-F:" '{print$NF}')
    done

    # Remove the stuff that we don't need
    rm -rf /etc/gitlab/gitlab-secrets.json \
           /var/lib/www

    # Remove the landing page...
    echo "Removing the landing page..."
    ( apt-get -qqy purge lighttpd >> /dev/null
      apt-get -qqy autoremove --purge >> /dev/null) &

    echo "Running 'gitlab-ctl reconfigure', this will take a minute..."
    (gitlab-ctl reconfigure 2>&1) >  /var/log/gitlab_reconfigure.log
    echo "  wrote log to /var/log/gitlab_reconfigure.log"

    # Set a randomized password.
    echo "Setting the password..."
    (gitlab-rake \
         gitlab:setup \
            RAILS_ENV=production \
            GITLAB_ROOT_PASSWORD=${gitlab_pass} \
            force=yes 2>&1 ) > /var/log/gitlab_set_pass.log
    echo "  wrote log to /var/log/gitlab_set_pass.log"

    cat <<EOC
-----------------------------------------------------------------------------

You can access GitLab via:
    Web URL:  http://${myip}
    User:     root
    Password: ${gitlab_pass}

For more information about this 1-Click, see: http://do.co/gitlabapp

To enable LetsEncrypt TLS certificates, see: http://do.co/le-gitlab

Happy Coding!
-----------------------------------------------------------------------------

EOC

    cp /etc/skel/.bashrc /root

else

   cat <<EOC
-----------------------------------------------------------------------------

To remove this message, run:
   cp /etc/skel/.bashrc /root

Please note that lighttpd is running to serve up the public splash page.
You may remove it via:
   apt-get -qqy purge lighttpd
   rm -rf /var/wwww

If you would like to let this script do the work for you, edit/change
/etc/gitlab/gitlab.rb.digitalocean and then run:
   bash $(readlink -f ${0})

Happy Coding!
-----------------------------------------------------------------------------
EOC

fi
```

**Comment by hugopeixoto** [comment]
I'll start working on changing the documentation to remove any references to the dokku web installer, and instead pointing them to configure the initial settings via the CLI.

Even if we end up with a splash page, the documentation should point users to the CLI.

**Comment by josegonzalez** [comment]
Yeah definitely. Would you also be interested in working on the actual work plan? This would be a bc-incompatible break, but worth it for us :)

**Comment by hugopeixoto** [comment]
> Yeah definitely. Would you also be interested in working on the actual work plan? This would be a bc-incompatible break, but worth it for us :)

If you mean removing the web installer, I'm up for it. I don't know what else still needs to be done. I think I saw some code related to the motd already.

**Comment by hugopeixoto** [comment]
I have created a PR that removes the web installer:

https://github.com/dokku/dokku/pull/4079

**Comment by hugopeixoto** [comment]
Is this the final plan?

https://github.com/dokku/dokku/issues/2247#issuecomment-229830997

> Ensure the deb prompts during interactive setup

I think this is already done. Debian did prompt me for the ssh key path in one of the installations.

Or do you think we should add the splash page?

**Comment by josegonzalez** [comment]
We should also add an motd and do the same sort of splash screen as Gitlab does (link here: https://github.com/dokku/dokku/issues/2247#issuecomment-275320176).

**Comment by nerg4l** [comment]
Is it still being worked on?

**Comment by josegonzalez** [comment]
Closing as there is a pull request open.
